# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsLayoutUtils.cpp
# Commit: 8079db41c325
# Full Hash: 8079db41c32526a9535056ce616c5383dc2a56c1
# Author: Timothy Nikkel <tnikkel@gmail.com>
# Date: 2022-08-29 21:45:26
# Regressor Bug: 1754436
# File Overlap Count: 3
# Description:
#   Bug 1787684. Only return views with widgets from FindFloatingViewContaining. r=emilio
#   
#   The code that bug 1754436 deleted checked to make sure there was a widget for the returned view, but it looks like it's possible in some edge cases to have an open popup that doesn't have a widget (most likely just for a split second). We need a widget to dispatch the event, and if it doesn't have a widget it's not visible so no point in touching it at all anyways.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D155819
# ==============================================================================

diff -r 251d145c0b4e -r 8079db41c325 layout/base/nsLayoutUtils.cpp
--- a/layout/base/nsLayoutUtils.cpp	Mon Aug 29 09:46:13 2022 +0000
+++ b/layout/base/nsLayoutUtils.cpp	Mon Aug 29 09:58:11 2022 +0000
@@ -1774,7 +1774,8 @@
 
 nsIFrame* nsLayoutUtils::GetPopupFrameForPoint(
     nsPresContext* aRootPresContext, nsIWidget* aWidget,
-    const mozilla::LayoutDeviceIntPoint& aPoint) {
+    const mozilla::LayoutDeviceIntPoint& aPoint,
+    GetPopupFrameForPointFlags aFlags /* = GetPopupFrameForPointFlags(0) */) {
   nsXULPopupManager* pm = nsXULPopupManager::GetInstance();
   if (!pm) {
     return nullptr;
@@ -1783,11 +1784,19 @@
   pm->GetVisiblePopups(popups);
   // Search from top to bottom
   for (nsIFrame* popup : popups) {
-    if (popup->PresContext()->GetRootPresContext() == aRootPresContext &&
-        popup->ScrollableOverflowRect().Contains(GetEventCoordinatesRelativeTo(
+    if (popup->PresContext()->GetRootPresContext() != aRootPresContext) {
+      continue;
+    }
+    if (!popup->ScrollableOverflowRect().Contains(GetEventCoordinatesRelativeTo(
             aWidget, aPoint, RelativeTo{popup}))) {
-      return popup;
-    }
+      continue;
+    }
+    if (aFlags & GetPopupFrameForPointFlags::OnlyReturnFramesWithWidgets) {
+      if (!popup->HasView() || !popup->GetView()->HasWidget()) {
+        continue;
+      }
+    }
+    return popup;
   }
   return nullptr;
 }